File: one_to_n.f90
program one_to_n
use omp_lib
use omp_lib_kinds
implicit none

integer,parameter                                 :: dp= selected_real_kind(15,300)
integer,parameter                                 :: long_int= selected_int_kind(16)
integer (kind=long_int)                           :: i,N=100000000, total_sum=0,threads
integer (kind=long_int),dimension(:),allocatable  :: array_N
real(kind=dp)                                     :: start_time=0.0_dp, end_time=0.0_dp

allocate(array_N(N))

!set each element to its index
do i=1,N
    array_N(i)=i
enddo

!sum each element to its index in parallel

!$OMP parallel private(i) shared (N,array_N) reduction(+:total_sum)
start_time = omp_get_wtime()
threads = omp_get_num_threads()
!$OMP do
do i=1,N
    total_sum=total_sum + array_N(i)
enddo
!$OMP end do
end_time = omp_get_wtime()
!$omp end parallel

print*,"Total sum: ", total_sum
print*,"Time: ", end_time-start_time, "seconds"
print*,"Threads: ", threads

deallocate(array_N)
endprogram

File: one_to_n.cpp
#include <omp.h>
#include <cstdio>
#include <vector>

int main(){
    
    long threads =-1;
    unsigned long i =0;
    long N=100000000; 
    long total_sum=0;
    double start_time=0.0;
    double end_time=0.0;
    
    std::vector<double> array_N;
    
    array_N.reserve(N);
    array_N.resize(N);
    
    //Set each array elem,ent to its index +1
    for(unsigned long j=0; j<array_N.size(); ++j){
        array_N[j]=j+1;
    }
    
    //sum up in parallel, with timing
    #pragma omp parallel private(i) shared(N,array_N) reduction(+:total_sum)
    {
        start_time = omp_get_wtime();
        threads = omp_get_num_threads();
        #pragma omp for 
            for(i=0; i<array_N.size(); ++i){
                    total_sum+=array_N[i];
                }
        end_time = omp_get_wtime();
    }
    
    printf("Total sum: %li \n",total_sum);
    printf("Time: %f seconds \n",(end_time-start_time));
    printf("Threads: %li \n",threads);
}
